motionplan/tpSpaceRRT.go

```diff
@@ -0,0 +1,435 @@
+//go:build !windows
+
+package motionplan
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/motionplan/tpspace"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+var (
+	defaultGoalCheck = 5
+	defaultAutoBB    = 1. // Automatic bounding box on driveable area as a multiple of start-goal distance
+
+	// whether to add intermediate waypoints.
+	defaultAddInt = false
+	// Add a subnode every this many mm along a valid trajectory. Large values run faster, small gives better paths
+	// Meaningless if the above is false.
+	defaultAddNodeEvery = 50.
+	defaultSearchRadius = 50.
+)
+
+// newCBiRRTMotionPlannerWithSeed creates a cBiRRTMotionPlanner object with a user specified random seed.
+func newTPSpaceMotionPlanner(
+	frame referenceframe.Frame,
+	seed *rand.Rand,
+	logger golog.Logger,
+	opt *plannerOptions,
+) (motionPlanner, error) {
+	if opt == nil {
+		return nil, errNoPlannerOptions
+	}
+	mp, err := newPlanner(frame, seed, logger, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tpspaceRRTMotionPlanner{
+		planner:   mp,
+		goalCheck: defaultGoalCheck,
+		autoBB:    defaultAutoBB,
+
+		addNodeEvery: defaultAddNodeEvery,
+
+		searchRadius:    defaultSearchRadius,
+		addIntermediate: defaultAddInt,
+	}, nil
+}
+
+// tpspaceRRTMotionPlanner.
+type tpspaceRRTMotionPlanner struct {
+	*planner
+	goalCheck int // Check if goal is reachable every this many iters
+
+	// size of bounding box within which to randomly sample points
+	// TODO: base this on frame limits?
+	autoBB       float64
+	addNodeEvery float64
+	searchRadius float64
+
+	addIntermediate bool
+}
+
+type candidate struct {
+	dist      float64
+	treeNode  *configurationNode
+	newNode   *configurationNode
+	finalGoal bool
+	err       error
+}
+
+// TODO: seed is not immediately useful for TP-space.
+func (mp *tpspaceRRTMotionPlanner) plan(ctx context.Context,
+	goal spatialmath.Pose,
+	seed []referenceframe.Input,
+) ([][]referenceframe.Input, error) {
+	solutionChan := make(chan *rrtPlanReturn, 1)
+
+	seedPos := spatialmath.NewZeroPose()
+
+	startNode := &configurationNode{endConfig: seedPos}
+	goalNode := &configurationNode{endConfig: goal}
+
+	utils.PanicCapturingGo(func() {
+		mp.rrtBackgroundRunner(ctx, seed, &rrtParallelPlannerShared{
+			&rrtMaps{
+				startMap: map[node]node{startNode: nil},
+				goalMap:  map[node]node{goalNode: nil},
+			},
+			nil,
+			solutionChan,
+		})
+	})
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case plan := <-solutionChan:
+		if plan != nil {
+			return plan.toInputs(), plan.err()
```

Comment:
Where is this implemented? How do we include the trajectory type in the inputs?

Reply:
Implemented here: https://github.com/viamrobotics/rdk/blob/main/motionplan/rrt.go#L44

This particular bit is just extracting/unwrapping the values from earlier.

The frame handed to the planner from the base is where the trajectories are stored, and part of the input is the indexes of the trajectories as provided by the base. So a single set of inputs would be something like "travel 345mm down the 47th trajectory of the 3rd trajectory family provided by this frame", which can be looked up by anything else with access to the frame (the base).

Reply:
Thank you!